### ISSUE TITLE
[Feature req] Allow building appimages (Useful if the app cant be redistributed)

### ISSUE THREAD

**Issue by Twig6943** [ivan-hc/AM#1243]
FDM cant't be distributed due to its license. That being said there's a repo with a pkg2appimage recipe https://github.com/gauthamp10/freedownloadmanager-linux-appimage

![image](https://github.com/user-attachments/assets/ff4c7e73-aa0e-48e4-9c8e-3116b77da538)

I want am to auto run this script (The script itself seems to be outdated but you probably get the idea)

![image](https://github.com/user-attachments/assets/525573e6-42e0-44fb-8462-265d3a95e6da)

Idk if this'll be added considering you've said that you dont like to fix others' mistakes before

**Comment by Twig6943** [comment]
Could u make a script for FDM then?

**Comment by ivan-hc** [comment]
This is a request to add an app.

**Comment by Twig6943** [comment]
> This is a request to add an app.

Should I create a new issue?

**Comment by ivan-hc** [comment]
> Should I create a new issue?

Given the magnitude of this request, you should at least test and document the creation of such an app. Adding assembly scripts to this repository is like delegating application maintenance to me.

**Comment by Twig6943** [comment]
> > Should I create a new issue?
> 
> Given the magnitude of this request, you should at least test and document the creation of such an app. Adding assembly scripts to this repository is like delegating application maintenance to me.

Working on it right now

**Comment by ivan-hc** [comment]
also, just noticed that instructions on the README are wrong and outdated, the `wget` command does not downloads pkg2appimage

**Comment by Twig6943** [comment]
> also, just noticed that instructions on the README are wrong and outdated, the `wget` command does not downloads pkg2appimage

Ye ye I've noticed that I'll make an updated repo

**Comment by ivan-hc** [comment]
> Ye ye I've noticed that I'll make an updated repo

it is not necessary, just create a script to build it and test it on your PC

**Comment by Twig6943** [comment]
Ok soo

1.The pkg2appimage repo seems to be incredibly outdated (Cant get a working pkg even with working pkg2appimage.appimage link

2.Cant get it to build with archimage -b freedownloadmanager

3.Idk how to use the "-t" option's build-on the fly option at all

**Comment by ivan-hc** [comment]
I'm already on the script

**Comment by Twig6943** [comment]
Here's the pkgbuild https://aur.archlinux.org/cgit/aur.git/tree/PKGBUILD?h=freedownloadmanager (Doesnt seem that long)

**Comment by ivan-hc** [comment]
Test this into an empty directory
```
#!/bin/sh

APP=freedownloadmanager

# DOWNLOADING THE DEPENDENCIES
if test -f ./appimagetool; then
	echo " appimagetool already exists" 1> /dev/null
else
	echo " Downloading appimagetool..."
	wget -q https://github.com/AppImage/appimagetool/releases/download/continuous/appimagetool-x86_64.AppImage -O appimagetool
fi
if test -f ./pkg2appimage; then
	echo " pkg2appimage already exists" 1> /dev/null
else
	echo " Downloading pkg2appimage..."
	wget -q https://github.com/ivan-hc/AM/raw/refs/heads/main/tools/pkg2appimage -O pkg2appimage
fi
chmod a+x ./appimagetool ./pkg2appimage

# CREATING THE HEAD OF THE RECIPE
echo "app: freedownloadmanager
binpatch: true

ingredients:
  script:
    - wget -q -c \"https://dn3.freedownloadmanager.org/6/latest/freedownloadmanager.deb\"


script:
  - mv opt/freedownloadmanager/* usr/bin/
  - find . -name icon.png -exec mv {} icon.png \;
  - cat > freedownloadmanager.desktop <<EOF
  - [Desktop Entry]
  - Type=Application
  - Name=freedownloadmanager
  - Icon=icon.png
  - Exec=fdm %u
  - Categories=Network;
  - Comment=Free Download Manager
  - EOF" > recipe.yml

# DOWNLOAD ALL THE NEEDED PACKAGES AND COMPILE THE APPDIR
echo "Compilling $APP from recipe using pkg2appimage..."
./pkg2appimage ./recipe.yml >/dev/null 2>&1

# EXPORT THE APP TO AN APPIMAGE
printf '#!/bin/sh\nexit 0' > ./desktop-file-validate # hack due to https://github.com/AppImage/appimagetool/pull/47
chmod a+x ./desktop-file-validate
PATH="$PATH:$PWD" ARCH=x86_64 ./appimagetool -n ./"$APP"/"$APP".AppDir 2>&1 | grep "Architecture\|Creating\|====\|Exportable"
if ! test -f ./*.AppImage; then
	echo "No AppImage available."; exit 1
fi
```

**Comment by ivan-hc** [comment]
the appimoge does not work btw

**Comment by ivan-hc** [comment]
Same result if I use the upstream pkg2appimage (mine does not creates the AppImage and allows appimagetool doing this, not to do this twice and to get rid of libfuse2)
```
#!/bin/sh

APP=freedownloadmanager

# DOWNLOADING THE DEPENDENCIES
if test -f ./appimagetool; then
	echo " appimagetool already exists" 1> /dev/null
else
	echo " Downloading appimagetool..."
	wget -q https://github.com/AppImage/appimagetool/releases/download/continuous/appimagetool-x86_64.AppImage -O appimagetool
fi
if test -f ./pkg2appimage; then
	echo " pkg2appimage already exists" 1> /dev/null
else
	echo " Downloading pkg2appimage..."
	wget -q https://github.com/AppImageCommunity/pkg2appimage/releases/download/continuous/pkg2appimage--x86_64.AppImage -O pkg2appimage
fi
chmod a+x ./appimagetool ./pkg2appimage

# CREATING THE HEAD OF THE RECIPE
echo "app: freedownloadmanager
binpatch: true

ingredients:
  script:
    - wget -q -c \"https://dn3.freedownloadmanager.org/6/latest/freedownloadmanager.deb\"


script:
  - mv opt/freedownloadmanager/* usr/bin/
  - find . -name icon.png -exec mv {} icon.png \;
  - cat > freedownloadmanager.desktop <<EOF
  - [Desktop Entry]
  - Type=Application
  - Name=freedownloadmanager
  - Icon=icon.png
  - Exec=fdm %u
  - Categories=Network;
  - Comment=Free Download Manager
  - EOF" > recipe.yml

# DOWNLOAD ALL THE NEEDED PACKAGES AND COMPILE THE APPDIR
echo "Compilling $APP from recipe using pkg2appimage..."
./pkg2appimage ./recipe.yml >/dev/null 2>&1

# EXPORT THE APP TO AN APPIMAGE
printf '#!/bin/sh\nexit 0' > ./desktop-file-validate # hack due to https://github.com/AppImage/appimagetool/pull/47
chmod a+x ./desktop-file-validate
PATH="$PATH:$PWD" ARCH=x86_64 ./appimagetool -n ./"$APP"/"$APP".AppDir 2>&1 | grep "Architecture\|Creating\|====\|Exportable"
if ! test -f ./*.AppImage; then
	echo "No AppImage available."; exit 1
else
	rm -f ./appimagetool ./pkg2appimage
fi
```

**Comment by Twig6943** [comment]
> Test this into an empty directory
> 
> ```
> #!/bin/sh
> 
> APP=freedownloadmanager
> 
> # DOWNLOADING THE DEPENDENCIES
> if test -f ./appimagetool; then
> 	echo " appimagetool already exists" 1> /dev/null
> else
> 	echo " Downloading appimagetool..."
> 	wget -q https://github.com/AppImage/appimagetool/releases/download/continuous/appimagetool-x86_64.AppImage -O appimagetool
> fi
> if test -f ./pkg2appimage; then
> 	echo " pkg2appimage already exists" 1> /dev/null
> else
> 	echo " Downloading pkg2appimage..."
> 	wget -q https://github.com/ivan-hc/AM/raw/refs/heads/main/tools/pkg2appimage -O pkg2appimage
> fi
> chmod a+x ./appimagetool ./pkg2appimage
> 
> # CREATING THE HEAD OF THE RECIPE
> echo "app: freedownloadmanager
> binpatch: true
> 
> ingredients:
>   script:
>     - wget -q -c \"https://dn3.freedownloadmanager.org/6/latest/freedownloadmanager.deb\"
> 
> 
> script:
>   - mv opt/freedownloadmanager/* usr/bin/
>   - find . -name icon.png -exec mv {} icon.png \;
>   - cat > freedownloadmanager.desktop <<EOF
>   - [Desktop Entry]
>   - Type=Application
>   - Name=freedownloadmanager
>   - Icon=icon.png
>   - Exec=fdm %u
>   - Categories=Network;
>   - Comment=Free Download Manager
>   - EOF" > recipe.yml
> 
> # DOWNLOAD ALL THE NEEDED PACKAGES AND COMPILE THE APPDIR
> echo "Compilling $APP from recipe using pkg2appimage..."
> ./pkg2appimage ./recipe.yml >/dev/null 2>&1
> 
> # EXPORT THE APP TO AN APPIMAGE
> printf '#!/bin/sh\nexit 0' > ./desktop-file-validate # hack due to https://github.com/AppImage/appimagetool/pull/47
> chmod a+x ./desktop-file-validate
> PATH="$PATH:$PWD" ARCH=x86_64 ./appimagetool -n ./"$APP"/"$APP".AppDir 2>&1 | grep "Architecture\|Creating\|====\|Exportable"
> if ! test -f ./*.AppImage; then
> 	echo "No AppImage available."; exit 1
> fi
> ```

Yep this doesn't work indeed 
![image](https://github.com/user-attachments/assets/02a6ba27-a8f3-4d7b-bbce-673f0c674c40)

**Comment by ivan-hc** [comment]
That said, good luck in your attempts to create it.

**Comment by Twig6943** [comment]
What about a portable pkg extracted from the .deb? @ivan-hc  (This does work)

**Comment by Twig6943** [comment]
Pls see
https://youtu.be/uYQ3P9DmwMw

**Comment by ivan-hc** [comment]
> Pls see https://youtu.be/uYQ3P9DmwMw

watching the video and every time you turned the cursor, you made me turn something else that I can't say.

PS: I'm already testing something for another similar app.

**Comment by Twig6943** [comment]
> Pls see https://youtu.be/uYQ3P9DmwMw

TLDR am -i fdm should;

1.Download  with `wget https://files2.freedownloadmanager.org/6/latest/freedownloadmanager.deb`

2.Extract it via something like;
```
ar x freedownloadmanager.deb
tar -xf data.tar.xst
```

3.Create .desktop file

[Desktop Entry]
Name=Free Download Manager
Comment=FDM is a powerful modern download accelerator and organizer
Keywords=download;manager;free;fdm;
Exec=fdm
Terminal=false
Type=Application
Icon=/opt/freedownloadmanager/icon.png
Categories=Network;FileTransfer;P2P;GTK;
StartupNotify=true

**Comment by ivan-hc** [comment]
Yo wrote an installation script manually and you don't noticed that.

If it is so easy for you creating installation scripts, why you always tell me to do stuff for you?

You have the time to play with all this. Please, don't say that it is not so!

**Comment by Twig6943** [comment]
Cuz most of the time I give up while trying to write the script

**Comment by ivan-hc** [comment]
> Cuz most of the time I give up while trying to write the script

If I had given up 3 years ago, I wouldn't have written "AM".

You give up too easily.

**Comment by Twig6943** [comment]
> You give up too easily.

Well that is kinda correct.. but anyways I do not know how to fetch this app's version

**Comment by ivan-hc** [comment]
> but anyways I do not know how to fetch this app's version (Afaik this is required for "am -u" )

You have the solution on the page where you downloaded that .deb file

I won't tell you anything else, you have to figure it out yourself.

**Comment by Twig6943** [comment]
1.wget (dl the .deb)
2.ar (extract the deb)
3.tar (extract the .tar.zst inside the .deb)
4.echo (create .desktop file)

but yeah I just cant pull off the update part (I need to make it curl the page and then look for the version part using cat and grep but yeah thats beyond my scripting knowledge) (I don't expect you to do stuff for me just stating the situation)

**Comment by ivan-hc** [comment]
You don't need `cat` if you use `curl` to read the web page. I already teached this to you a while ago.

A thing I can suggest is to analyze the web page from your browser, so you know what to `grep`

**Comment by Twig6943** [comment]
```
#!/bin/sh

# AM INSTALL SCRIPT VERSION 3.5
set -u
APP=fdm
SITE="freedownloadmanager.org"

# CREATE DIRECTORIES AND ADD REMOVER
[ -n "$APP" ] && mkdir -p "/opt/$APP/tmp" "/opt/$APP/icons" && cd "/opt/$APP/tmp" || exit 1
printf "#!/bin/sh\nset -e\nrm -f /usr/local/bin/$APP\nrm -R -f /opt/$APP" > ../remove
printf '\n%s' "rm -f /usr/local/share/applications/$APP-AM.desktop" >> ../remove
chmod a+x ../remove || exit 1

# DOWNLOAD AND PREPARE THE APP, $version is also used for updates
version=$(curl -Ls https://www.freedownloadmanager.org/download-fdm-for-linux.htm | grep -oP 'FDM \K[0-9]+\.[0-9]+\.[0-9]+' | head -1)
[ -n "$version" ] || { echo "Error fetching version"; exit 1; }
wget -q "https://files2.freedownloadmanager.org/6/latest/freedownloadmanager.deb" || exit 1
ar x freedownloadmanager.deb || exit 1
tar -xf data.tar.zst || exit 1
mv ./opt/freedownloadmanager/* "/opt/$APP" || exit 1
rm -R -f ./tmp || exit 1
echo "$version" > "/opt/$APP/version"
chmod a+x "/opt/$APP/fdm" || exit 1

# LINK TO PATH
ln -s "/opt/$APP/fdm" "/usr/local/bin/$APP" || exit 1

# SCRIPT TO UPDATE THE PROGRAM
cat > /opt/$APP/AM-updater << 'EOF'
#!/bin/sh
set -u
APP=freedownloadmanager
URL="https://www.freedownloadmanager.org/download-fdm-for-linux.htm"
version0=$(cat "/opt/$APP/version")
version=$(curl -Ls "$URL" | grep -oP 'FDM \K[0-9]+\.[0-9]+\.[0-9]+' | head -1)
[ -n "$version" ] || { echo "Error fetching version"; exit 1; }
if [ "$version" != "$version0" ]; then
    mkdir "/opt/$APP/tmp" && cd "/opt/$APP/tmp" || exit 1
    wget -q "https://files2.freedownloadmanager.org/6/latest/freedownloadmanager.deb" || exit 1
    ar x freedownloadmanager.deb || exit 1
    tar -xf data.tar.zst || exit 1
    mv --backup=t ./opt/freedownloadmanager/* "/opt/$APP" || exit 1
    chmod a+x "/opt/$APP/fdm" || exit 1
    echo "$version" > "/opt/$APP/version"
    rm -R -f ./tmp
    notify-send "$APP is updated to version $version!"
else
    echo "Update not needed!"
fi
EOF
chmod a+x /opt/$APP/AM-updater || exit 1

# ICON
wget -q "https://cdn.icon-icons.com/icons2/3053/PNG/512/free_download_manager_macos_bigsur_icon_190155.png" -O "/opt/$APP/icons/$APP" || exit 1

# LAUNCHER
echo "[Desktop Entry]
Name=Free Download Manager
Exec=$APP
Terminal=false
Icon=/opt/$APP/icons/$APP
StartupWMClass=freedownloadmanager
Type=Application
Categories=Application;Internet;" > "/usr/local/share/applications/$APP-AM.desktop" || exit 1
```

**Comment by ivan-hc** [comment]
I added it as an AppImage built on-the-fly.

As for the other requests, I'll let you solve them yourself.
